### PR TITLE
08-shopping_cart-cart_item-shopping_cart_management

### PR DESCRIPTION
**Pull request by andrey-kiyas**
Added:
- shopping cart
- cart item
- shopping cart management

### CODE REVIEW

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @Operation(summary = "Retrieve user's shopping cart",
+            description = "Retrieve user's shopping cart")
+    @GetMapping
+    @PreAuthorize("hasRole('USER')")
+    public ShoppingCartResponseDto getShoppingCart(Authentication authentication) {
+        String email = authentication.getName();
+        return shoppingCartService.getShoppingCart(email);
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/controller/ShoppingCartController.java]:
```suggestion
        User user = (User) authentication.getPrincipal();
        return shoppingCartService.getById(user.getId());
```

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    @Operation(summary = "Add book to the shopping cart",
+            description = "Add book to the shopping cart")
+    @PostMapping
+    @PreAuthorize("hasRole('USER')")
+    public ShoppingCartResponseDto addBookToShoppingCart(
+            Authentication authentication,
+            @RequestBody @Valid CartItemRequestDto cartItemRequestDto) {
+        String email = authentication.getName();
+        shoppingCartService.addCartItem(email, cartItemRequestDto);
+        return shoppingCartService.getShoppingCart(email);
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/controller/ShoppingCartController.java]:
```suggestion
        User user = (User) authentication.getPrincipal();
        return shoppingCartService.addToCart(request, user.getId());
```

File: src/main/java/com/onlinebookstore/mapper/ShoppingCartMapper.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import com.onlinebookstore.dto.shoppingcart.ShoppingCartResponseDto;
+import com.onlinebookstore.model.CartItem;
+import com.onlinebookstore.model.ShoppingCart;
+import java.util.HashSet;
+import java.util.Set;
+import org.mapstruct.AfterMapping;
+import org.mapstruct.Mapper;
+import org.mapstruct.MappingTarget;
+
+@Mapper(config = MapperConfig.class)
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/mapper/ShoppingCartMapper.java]:
```suggestion
@Mapper(config = MapperConfig.class, uses = CartItemMapper.class)
```

File: src/main/java/com/onlinebookstore/mapper/ShoppingCartMapper.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+public interface ShoppingCartMapper {
+    ShoppingCartResponseDto toDto(ShoppingCart shoppingCart);
+
+    ShoppingCart toEntity(ShoppingCartResponseDto shoppingCartResponseDto);
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/mapper/ShoppingCartMapper.java:17]:
```suggestion
    @Mapping(source = "userId", target = "user.id")
    @Mapping(source = "cartItems", target = "cartItems")
    ShoppingCart toEntity(ShoppingCartResponseDto shoppingCartResponseDto);
```

File: src/main/java/com/onlinebookstore/mapper/ShoppingCartMapper.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        Set<CartItemResponseDto> cartItemSet = new HashSet<>();
+        for (CartItem item : shoppingCart.getCartItems()) {
+            CartItemResponseDto cartItem = new CartItemResponseDto();
+            cartItem.setId(item.getId());
+            cartItem.setBookId(item.getBook().getId());
+            cartItem.setBookTitle(item.getBook().getTitle());
+            cartItem.setQuantity(item.getQuantity());
+            cartItemSet.add(cartItem);
+        }
+        shoppingCartResponseDto.setCartItems(cartItemSet);
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/mapper/ShoppingCartMapper.java]:
```suggestion
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/com/onlinebookstore/mapper/ShoppingCartMapper.java (hunk 5 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/mapper/ShoppingCartMapper.java]: Agree. This decision wasn't correct. Corrected by means of mapping.

File: src/main/java/com/onlinebookstore/mapper/ShoppingCartMapper.java (hunk 4 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/mapper/ShoppingCartMapper.java:17]: Thanks, fixed it.

File: src/main/java/com/onlinebookstore/mapper/ShoppingCartMapper.java (hunk 3 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/mapper/ShoppingCartMapper.java]: Thanks, fixed it.

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java (hunk 2 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/controller/ShoppingCartController.java]: I can't agree with the rationale behind this change. Login (email) is in this case a constant value, from which I initially wanted to build on the issue of user search. Moreover, rewriting the code for searching by id gives an increase in the amount of code. In addition, an indexed email field (unique) gives the same result as when working with id.

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java (hunk 1 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/controller/ShoppingCartController.java]: I can't agree with the rationale behind this change. Login (email) is in this case a constant value, from which I initially wanted to build on the issue of user search. Moreover, rewriting the code for searching by id gives an increase in the amount of code. In addition, an indexed email field (unique) gives the same result as when working with id.

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java (hunk 2 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/controller/ShoppingCartController.java]: Thanks, fixed it.

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java (hunk 1 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/controller/ShoppingCartController.java]: Thanks, fixed it.

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    @Operation(summary = "Update quantity of a book in the shopping cart",
+            description = "Update quantity of a book in the shopping cart")
+    @PutMapping("cart-items/{cartItemId}")
+    @PreAuthorize("hasRole('USER')")
+    public ShoppingCartResponseDto updateCartItem(
+            Authentication authentication,
+            @PathVariable Long cartItemId,
+            @RequestBody @Valid CartItemRequestDto requestDto) {
+        User user = (User) authentication.getPrincipal();
+        String email = authentication.getName();
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/controller/ShoppingCartController.java]:
```suggestion
  
```

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+    @Operation(summary = "Remove a book from the shopping cart",
+            description = "Remove a book from the shopping cart")
+    @DeleteMapping("cart-items/{cartItemId}")
+    @PreAuthorize("hasRole('USER')")
+    public ShoppingCartResponseDto deleteCartItem(Authentication authentication,
+                                                  @PathVariable Long cartItemId) {
+        User user = (User) authentication.getPrincipal();
+        String email = authentication.getName();
+        shoppingCartService.deleteCartItem(user.getId(), cartItemId);
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/controller/ShoppingCartController.java]:
```suggestion
        User user = (User) authentication.getPrincipal();
        shoppingCartService.deleteCartItem(user.getId(), cartItemId);
```

File: src/main/java/com/onlinebookstore/dto/cartitem/CartItemRequestDto.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    @Positive
+    private Long bookId;
+    @Positive
+    private int quantity;
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/dto/cartitem/CartItemRequestDto.java]: add @NotNull

File: src/main/java/com/onlinebookstore/dto/cartitem/CartItemUpdateDto.java [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package com.onlinebookstore.dto.cartitem;
+
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class CartItemUpdateDto {
+    @Positive
+    private int quantity;
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/dto/cartitem/CartItemUpdateDto.java:11]: add @NotNull

File: src/main/java/com/onlinebookstore/mapper/CartItemMapper.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    CartItem toEntity(CartItemUpdateDto cartItemUpdateDto);
+
+    @Mappings({
+            @Mapping(target = "bookId", source = "book.id"),
+            @Mapping(target = "bookTitle", source = "book.title")
+    })
+    CartItemResponseDto toDto(CartItem cartItem);
+
+    @AfterMapping
+    default void setBookId(@MappingTarget CartItem cartItem,
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/mapper/CartItemMapper.java]:
```suggestion
    default void setBook(@MappingTarget CartItem cartItem,
```

File: src/main/java/com/onlinebookstore/service/ShoppingCartService.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
+package com.onlinebookstore.service;
+
+import com.onlinebookstore.dto.cartitem.CartItemRequestDto;
+import com.onlinebookstore.dto.shoppingcart.ShoppingCartResponseDto;
+import com.onlinebookstore.model.ShoppingCart;
+
+public interface ShoppingCartService {
+    ShoppingCart getShoppingCartById(Long id);
+
+    ShoppingCartResponseDto getShoppingCartDtoByUserId(Long id);
```
Comment (Rommelua) [inline src/main/java/com/onlinebookstore/service/ShoppingCartService.java]:
```suggestion
    ShoppingCartResponseDto getShoppingCartById(Long id);
```

**Review by Rommelua — APPROVED** [review]

File: src/main/java/com/onlinebookstore/service/ShoppingCartService.java (hunk 11 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/service/ShoppingCartService.java]: Thanks, fixed it.

File: src/main/java/com/onlinebookstore/mapper/CartItemMapper.java (hunk 10 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/mapper/CartItemMapper.java]: Thanks, fixed it.

File: src/main/java/com/onlinebookstore/dto/cartitem/CartItemUpdateDto.java (hunk 9 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/dto/cartitem/CartItemUpdateDto.java:11]: Thanks, fixed it.

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java (hunk 7 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/controller/ShoppingCartController.java]: Thanks, fixed it.

File: src/main/java/com/onlinebookstore/dto/cartitem/CartItemRequestDto.java (hunk 8 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/dto/cartitem/CartItemRequestDto.java]: Thanks, fixed it.

File: src/main/java/com/onlinebookstore/controller/ShoppingCartController.java (hunk 6 above)
Comment (andrey-kiyas) [reply src/main/java/com/onlinebookstore/controller/ShoppingCartController.java]: Thanks, fixed it.

**Pull request merged**